radicle-tracker/src/metadata.rs

```diff
@@ -0,0 +1,240 @@
+use std::{
+    collections::{HashMap, HashSet},
+    convert::Infallible,
+    hash::Hash,
+    str::FromStr,
+};
+
+/// The metadata that is related to an issue.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Metadata<User: Eq + Hash> {
+    labels: HashSet<Label>,
+    assignees: Assignees<User>,
+}
+
+impl<User: Eq + Hash> Metadata<User> {
+    /// Initialise empty metadata.
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Metadata {
+            labels: HashSet::new(),
+            assignees: Assignees::new(),
+        }
+    }
+
+    /// Get a reference to the [`Label`]s in the metadata.
+    pub fn labels(&self) -> &HashSet<Label> {
+        &self.labels
+    }
+
+    /// Add a [`Label`] to the set of labels in the metadata.
+    pub fn add_label(&mut self, label: Label) -> bool {
+        self.labels.insert(label)
+    }
+
+    /// Remove a [`Label`] from the set of labels in the metadata.
+    pub fn remove_label(&mut self, label: &Label) -> bool {
+        self.labels.remove(label)
+    }
+
+    /// Get a reference to the [`Assignees`] in the metadata.
+    pub fn assignees(&self) -> &Assignees<User> {
+        &self.assignees
+    }
+
+    /// Add a `User` to the set of [`Assignees`] in the metadata.
+    pub fn add_assignee(&mut self, assignee: User) -> bool {
+        self.assignees.add(assignee)
+    }
+
+    /// Remove a `User` from the set of [`Assignees`] in the metadata.
+    pub fn remove_assignee(&mut self, assignee: &User) -> bool {
+        self.assignees.remove(assignee)
+    }
+}
+
+/// The title of an issue.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Title(String);
+
+impl Title {
+    /// Create a new title.
+    pub fn new(title: String) -> Self {
+        Title(title)
+    }
+}
+
+impl std::ops::Deref for Title {
+    type Target = String;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<&str> for Title {
+    fn from(s: &str) -> Self {
+        Title(s.to_string())
+    }
+}
+
+impl FromStr for Title {
+    type Err = Infallible;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Title(s.to_string()))
+    }
+}
+
+/// A comment of an issue is composed of its [`Comment::author`], the
+/// [`Comment::content`] of the comment, and its [`Comment::reactions`].
+///
+/// It has a unique identifier (of type `Id`) chosen by the implementor.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Comment<Id, User: Eq + Hash> {
+    identifier: Id,
+    author: User,
+    content: String,
+    reactions: HashSet<Reaction<User>>,
```

Comment:
Timestamp maybe good to have

Reply:
Ya, that occurred to me too, thanks! Is there UTC time in Rust? I would imagine that's best, then the app can convert based on the user's local settings

Reply:
There's `SystemTime` which could be good enough for now :thinking:

Reply:
Please do _not_ use UTC but a proper zoned local time representation (I would assume the [time](https://docs.rs/time/0.2.7/time/) crate has one). UTC is subject to leap seconds, which are announced by committee decision like approx. 6 months in advance, and that will eventually cause trouble.

Reply:
Ah, ok thanks for the call out :)

Reply:
Would this be what we're looking for? https://docs.rs/time/0.2.7/time/struct.OffsetDateTime.html I haven't been bitten by time yet in my career so some guidance would help :sweat_smile:

Reply:
Ya so.. it turns out the Rust time libraries stuff their haggis with... well, you don’t want to look too closely. In particular, they don’t handle leap second nor time zone conversions correctly (this has to take DST into account, which they don’t).

Since the goal here is to merely _display_ completely unreliable timestamps in a halfway consistent manner in the _viewers_ local time zone, I’d amend my previous comment and suggest we type this in a sufficiently imprecise way such that the API doesn’t suggest you can do things with the timestamp which are wrong. Namely, the type shouldn’t implement `Ord` or `PartialOrd`, and should have a precision of no more than, say, 60secs. In other words, a newtype around `SystemTime::duration_since(UNIX_EPOCH)`, which can only be used to render a _relative_ time (“X units ago”).

Reply:
So just to clarify, you're suggesting that we add a newtype for something that we can represent as a date and time local to the user, but can't say use for ordering at all?

We'll also need some kind of `PartialOrd` timestamp field for when it comes to the merkl-dag and operation based stuff so that we have some total order to fall back on.